accepted/0000-link-packages-to-source-and-build.md

```diff
@@ -0,0 +1,514 @@
+
+Sigstore supports both email and CI/CD workload-based OIDC identities:
+
+##### Email based identity provider
+For instance, Google, GitHub, Microsoft (supported by Sigstore today) or any other IdP that Sigstore might support.
+
+**Pros**
+- Separate party for trust: it would be difficult for an attacker to compromise the repository and one of these IdPs.
+- Many (but not all!) users have these accounts already.
+- Portable: supports publishing from most environments (requiring a browser and an IdP login).
+- Minimal changes to existing publish workflow for maintainers that are publishing directly.
+- Reliability: these IdPs already work at scale and have full-time support.
+
+**Cons**
+- Privacy: the identities used to sign packages are public (though users can make a pseudonymous account).
+- Privacy: Only a handful of identity providers are supported today that might not suit all users (GitHub, Google or Microsoft).
+- Security: Some users might only have secondary, less secure, accounts on one of the supported services making it an easier target.
+- Does not include provenance information about where the release came from and how it was built.
+- The email might not match the one used on the registry making it hard to validate that it's from a legitimate maintainer.
+
+##### CI/CD workload identity provider
+For instance, Github Actions (supported by Sigstore today) or any other CI/CD system that can provide OIDC workload identities with required claims.
+
+**Pros**
+- Support for pseudonyms/privacy: does not require public identities other than what already corresponds to the package (e.g. repository).
+- Links the package to the source repository and includes information about and how it was built: commit SHA, workflow run id etc.
+- Vendor-neutral: supports any CI/CD system supporting OIDC workload identities.
+- OIDC workflow identities could also be used to authenticate the package publish (restricting publishing to a specific source code repo and removing the need for long-lived access tokens).
+
+**Cons**
+- Limited support today: Only GitHub Actions is fully supported in Fulcio today.
+- Friction: Requires publishing from supported CI/CD providers.
+- CI/CD providers will need to operate an IdP and get it integrated with Sigstore.
+
+Using workload identities along with short-lived x.509 certificates is currently the only solution that meets our key requirements:
+- Linking packages to the source and build.
+- Doesn't expose any personal identifiable information about maintainers.
+- Avoids developer-managed keys by placing trust in the CI/CD identity provider.
+
+Today this means only commercial CI/CD providers will be supported. This is far from where all developers are publishing today so this will add friction and slow adoption. All major CI/CD providers offer free plans for Open Source projects and we hope that all packages will eventually be built out in the open to make supply chain attacks harder to execute.
```

Comment:
If this is the goal, you are going to get questions about 2FA and automated publishing. While strides have been made, are there further efforts to continue to make publishing from CI a reasonable experience given that this will effectively force _all_ maintainers of large-scale packages to publish from CI?

Reply:
Copying answer from another thread. 

> are there further efforts to continue to make publishing from CI a reasonable experience given that this will effectively force all maintainers of large-scale packages to publish from CI?

Yes! We're working on a proposal which will allow authorising the publish using the same OIDC id token from the CI/CD system. So instead of using a long-lived access token to authorize the publish, you would log into your npm dashboard and set up which source repo is allowed to publish what package. The publish step then exchanges the OIDC id token for a short lived access token after the npm registry has checked the source repo from the id token matches the allowed one.

Once you've set up publishing in this way, you would not need to use long-lived access tokens and the risk of stolen credentials basically goes away if you also use 2FA on your npm account. The setup for this would be exactly the same as what we're proposing in this RFC so you essentially get source and build linking for free if you set up OIDC publishing. I imagine adoption would pick up a lot once we have both features in place and all major CI/CD providers are fully supported.

I think we can also ease the migration by providing reusable workflows for publishing npm packages on CI/CD. On actions you could start by using a [manual workflow trigger](https://github.blog/changelog/2020-07-06-github-actions-manual-triggers-with-workflow_dispatch/) to publish.

Reply:
While longer term we are hoping to implement something on top of OIDC we are currently working on Granular access tokens as a successor to automation tokens. These tokens will:

* Be able to be scoped to a specific package or scope
* Scoped to read or read/write permissions
* Has an expiry date

As mentioned above longer term we are looking at OIDC to remove the needs for tokens and a streamlined GitHub experience to make setting up everything a breeze. All of this is intended to be build on open technologies so that other CI/CD environments can benefit from similar implementation and that we can explore scaling these approaches to other package registries.

None of the above is part of the signing work outlined in this RFC though.

What I can say is that the remainder of this year has a big focus on "package security" for publishing in the same way we've spent the last 9 months extremely focused on "account security"